File: python/src/core/services/extended_service_provider.py
from typing import Type, Any
from utils.service_container import ServiceCollection, ServiceProvider, ServiceNotFoundError


class ExtendedServiceProvider:
    def __init__(self, provider: 'ServiceProvider'):
        self._provider = provider
        self._collection = ServiceCollection()
        self._collection.add_scoped('ServiceProvider', lambda sp: self)

    @property
    def services(self) -> ServiceCollection:
        return self._collection

    def get_service(self, service_type: Type) -> Any:
        local_provider = self._collection.build()
        try:
            return local_provider.get_service(service_type) or self._provider.get_service(service_type)
        except ServiceNotFoundError:
            return self._provider.get_service(service_type)

File: core/sdk/dotnet/Services/ExtendedServiceProvider.cs
using Microsoft.Extensions.DependencyInjection;


namespace Agience.Core.Services
{
    public class ExtendedServiceProvider : IServiceProvider
    {
        public IServiceCollection Services => _collection;

        private readonly IServiceProvider _provider;
        private readonly IServiceCollection _collection = new ServiceCollection();

        public ExtendedServiceProvider(IServiceProvider provider)
        {
            _provider = provider;
            _collection.AddScoped<IServiceProvider>(sp => this);
        }

        public object? GetService(Type serviceType)
        {
            return _collection.BuildServiceProvider().GetService(serviceType) ?? _provider.GetService(serviceType);
        }        
    }
}
